### PR TITLE
8278021: Fix warnings in macOS glass native code and treat warnings as errors

### PR DESCRIPTION
**Pull request by beldenfox**
Turning on warnings-as-errors for the macOS glass native code. Deprecated declarations are excluded and still appear as warnings.

In the code that tries to locate the application's dock icon there were three instances where `NO` was being passed into a method that required a pointer to a `BOOL`, not a `BOOL`. I suspect the intent was to check that the path pointed to an existing file but not a directory. Since JavaFX has gone this long without screening out directories correctly I decided not to fix that behavior except at the very end.

The only other changes of note are sending some NSNotification objects to delegate API's that require them even though we know they're ignored on the other side. It was the easiest way to get rid of the warning.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8278021](https://bugs.openjdk.org/browse/JDK-8278021): Fix warnings in macOS glass native code and treat warnings as errors (**Bug** - P4)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/687/head:pull/687` \
`$ git checkout pull/687`

Update a local copy of the PR: \
`$ git checkout pull/687` \
`$ git pull https://git.openjdk.org/jfx.git pull/687/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 687`

View PR using the GUI difftool: \
`$ git pr show -t 687`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/687.diff">https://git.openjdk.org/jfx/pull/687.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/687#issuecomment-983890067)

### CODE REVIEW

**Review by overheadhunter — COMMENTED** [review]
What about usages of deprecated [NSWindowStyleMasks](https://developer.apple.com/documentation/appkit/nswindowstylemask) in `GlassWindow.m`? Shouldn't they get fixed as well?

e.g. `NSTexturedBackgroundWindowMask` → `NSWindowStyleMaskTexturedBackground`

**Comment by beldenfox** [comment]
Yes, the deprecated constants should be fixed as the code is updated. They weren't included in this PR because there's a lot of them and they're all cosmetic.

**Comment by kevinrushforth** [comment]
I'll review this. The large number of warnings can hide a real issue (which I noticed the other day while working on a bug).

**Comment by kevinrushforth** [comment]
@beldenfox I let this PR drop off  my radar a while ago. Sorry about that!

I've been taking a look at our various warnings in a number of places, and was recently reminded of this by a comment you made in PR #1351 :

> I also updated a few deprecated constants so this file no longer generates compiler warnings.

Would you be willing to revive this PR, provided of course, that I actually review it this time? It will likely need to be updated in light of other changes in the past year or so.

**Comment by beldenfox** [comment]
@kevinrushforth I'd love to revive this PR! The keyboard related bugs are settling down so I was planning to do some cleanup of the Mac code this cycle. I'll take another look at this.

**Comment by beldenfox** [comment]
Re-opening this PR. Instead of disabling deprecation warnings globally I added pragmas to the relevant files. This will make it easier to continue the clean-up; someone can file a bug against a specific file, remove the pragma, and fix it up without having to wade through the warnings generated by the other files.

**Comment by beldenfox** [comment]
The pre-submit error is due to a call to `[NSApp activate]` which is only available in macOS 14.0 and up. My local builds use a 14.x SDK but the pre-submit tests are built with 13.3. It's an easy fix but I first need to find out why the SDK matters; the minimum target is 11.0 so I should have seen a warning.

**Comment by beldenfox** [comment]
Ouch.

We're targeting a minimum macOS version of 11.0 but `[NSApp activate]` is only available on 14.0 and above. Xcode 15 should issue a warning about this but doesn't *if this is the only version warning in the file*. I've tried this with a standalone app and also tweaked the Glass code and the result is the same; I get no warning unless I add some other 14.0-specific call in the same file. At that point I'll get a warning about the other violation in addition to the `[NSApp activate]` call.

Adding an `@available` check around this is the right thing to do but not enough to appease the pre-submit bot which is building with the 13.3 SDK. To clean this up either the bot needs to be upgraded or I need to add a localized pragma to disable this check.

**Comment by kevinrushforth** [comment]
> We're targeting a minimum macOS version of 11.0 but `[NSApp activate]` is only available on 14.0 and above. Xcode 15 should issue a warning about this but doesn't _if this is the only version warning in the file_.

Sigh. I'm not much a fan of how Apple handles runtime version checks, and this seems like one more indication of that.

> Adding an `@available` check around this is the right thing to do but not enough to appease the pre-submit bot which is building with the 13.3 SDK.

Technically the `@available` check is redundant (the only place that `requiresActivation` is set to `YES` is already in an `@available` check), but I can see how the compiler wouldn't know that. Adding it around the use site seems fine; too bad it is insufficient.

> To clean this up either the bot needs to be upgraded

Definitely not. This matches our production build environment. We use Xcode 14.3.1, which includes `MacOSX13.sdk`.

> or I need to add a localized pragma to disable this check.

So basically, a pragma to workaround a limitation of Xcode. This might be the best option if it isn't too ugly.

**Comment by kevinrushforth** [comment]
Btw, I ran this PR through our CI and got the expected error:

```
Starting process 'command 'jfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang''. Working directory: jfx/modules/javafx.graphics Command: jfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-macos-11 -mmacosx-version-min=11.0 -isysroot jfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.sdk -iframeworkjfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.sdk/System/Library/Frameworks -arch x86_64 -I/Users/java_re/jenkins/workspace/jfx-submit/label/mac-13-x64/jfx/bootjdk/jdk-21.0.2.jdk/Contents/Home/include -I/Users/java_re/jenkins/workspace/jfx-submit/label/mac-13-x64/jfx/bootjdk/jdk-21.0.2.jdk/Contents/Home/include/darwin -std=c99 -c  -O3 -DNDEBUG -DGL_SILENCE_DEPRECATION -DMACOS_MIN_VERSION_MAJOR=11 -DMACOS_MIN_VERSION_MINOR=0 -Werror -Ijfx/modules/javafx.graphics/build/gensrc/headers/javafx.graphics -Ijfx/modules/javafx.graphics/src/main/native-glass/mac -Ijfx/modules/javafx.graphics/src/main/native-glass/mac/a11y -o jfx/modules/javafx.graphics/build/native/glass/mac/GlassApplication.obj jfx/modules/javafx.graphics/src/main/native-glass/mac/GlassApplication.m
Successfully started process 'command 'jfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang''
jfx/modules/javafx.graphics/src/main/native-glass/mac/GlassApplication.m:304:24: error: instance method '-activate' not found (return type defaults to 'id') [-Werror,-Wobjc-method-access]
                [NSApp activate];
                       ^~~~~~~~
jfx/buildSrc/build/build-tools/devkit-macosx-Xcode14.3.1+1.0.tar/Xcode/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSApplication.h:158:12: note: receiver is instance of class declared here
@interface NSApplication : NSResponder <NSUserInterfaceValidations, NSMenuItemValidation, NSAccessibilityElement, NSAccessibility>
           ^
1 error generated.
```

**Comment by beldenfox** [comment]
> So basically, a pragma to workaround a limitation of Xcode.

I wouldn't describe this as a limitation of Xcode. After all, the glass code is trying to access a 14.0 API using a 13.3 SDK. If the language was C++ the call wouldn't be in the headers and the code wouldn't compile.

If we launch a jfx22 app on macOS 13 in theory we're sending NSApp an `activate` selector that it doesn't recognize. Or maybe it does? It's possible that this is an old internal call that Apple finally documented in 14.0. That would explain why they would be lenient about issuing a warning. I'll run some tests.

**Comment by kevinrushforth** [comment]
> > So basically, a pragma to workaround a limitation of Xcode.
> 
> I wouldn't describe this as a limitation of Xcode. After all, the glass code is trying to access a 14.0 API using a 13.3 SDK. If the language was C++ the call wouldn't be in the headers and the code wouldn't compile.

In C++ we would need to use reflection (dlsym), so I can see your point.

> If we launch a jfx22 app on macOS 13 in theory we're sending NSApp an `activate` selector that it doesn't recognize. Or maybe it does? It's possible that this is an old internal call that Apple finally documented in 14.0. That would explain why they would be lenient about issuing a warning. I'll run some tests.

No, we're not. The `@available` generates a runtime check so that code is executed only if the runtime platform is macOS 14.x or later.

**Comment by beldenfox** [comment]
> No, we're not. The `@available` generates a runtime check so that code is executed only if the runtime platform is macOS 14.x or later.

Right, forgot about the other `@available` check.

**Comment by beldenfox** [comment]
This is now compiling on 13.3 and I was able to do a Robot test run on macOS 14 with only one failure that's also present in the master branch. I still don't know why the compiler doesn't flag the `activate` call, in macOS 13 `NSApp` does not respond to that selector and will throw an exception if it's sent.

**Comment by kevinrushforth** [comment]
This now passes on our CI build systems. I'll review it next week some time. Most of the changes are trivial, but there are a couple I want to double-check that there is no change in behavior.

**Review by kevinrushforth — APPROVED** [review]
I ran a full set of headful tests on macOS 13 and 14.

I also ran some local tests and verified the modified code.

All looks good.

**Comment by kevinrushforth** [comment]
@aghaisas said he could be the second reviewer. Thanks, Ajit.
